glusterd2/commands/volumes/volume-profile.go

```diff
@@ -0,0 +1,309 @@
+}
+
+// Calculate Percentage latency for each fop in cumulative/interval stats
+func calculatePercentageLatencyForEachFop(stats StatType) StatType {
+	// Calculate sum of (hits * avgLatency) of all fops in Cumulative Stats. Used Later to calculate
+	// %-Latency for each FOP.
+	var tmpPercentageAvgLatency float64
+	for key := range stats.StatsInfo {
+		tmpAvgLatency, _ := strconv.ParseFloat(stats.StatsInfo[key]["avglatency"], 64)
+		tmpHits, _ := strconv.ParseFloat(stats.StatsInfo[key]["hits"], 64)
+		tmpPercentageAvgLatency += tmpAvgLatency * tmpHits
+	}
+	// Calculate %-Latency for each fop
+	// %-Avg-Latency  for one fop = 100 * (Hits for that fop * AvgLatency for that fop ) / (sum of hits * avgLatency of all fop)
+	for key := range stats.StatsInfo {
+		tmpAvgLatency, _ := strconv.ParseFloat(stats.StatsInfo[key]["avglatency"], 64)
+		tmpHits, _ := strconv.ParseFloat(stats.StatsInfo[key]["hits"], 64)
+		tmpPercentageAvgLatencyForEachFop := 100 * ((tmpAvgLatency * tmpHits) / tmpPercentageAvgLatency)
+		stats.StatsInfo[key]["%-latency"] = fmt.Sprintf("%f", tmpPercentageAvgLatencyForEachFop)
+	}
+	stats.PercentageAvgLatency = tmpPercentageAvgLatency
+
+	return stats
+}
+
+// populateStatsWithFop populates both interval/cumulative stats with fops and its profile info
+func populateStatsWithFop(stats StatType, key string, value string, ktype keyType) StatType {
+	var fop string
+	var k string
+	// Decode the fop and stat for the given key
+	switch ktype {
+	case cumulativeType:
+		fop, k = decodeCumulativeKey(key)
+		break
+	case intervalType:
+		fop, k = decodeIntervalKey(key)
+		break
+	}
+	if fop != "" && fop != "NULL" {
+		if _, ok := stats.StatsInfo[fop]; ok {
+			stats.StatsInfo[fop][k] = value
+		} else {
+			// Create new map with fop as key if this particular Fop is encountered for the first time
+			stats.StatsInfo[fop] = make(map[string]string)
+			stats.StatsInfo[fop][k] = value
+		}
+	} else {
+		if k == "read" {
+			stats.DataRead = value
+		} else if k == "write" {
+			stats.DataWrite = value
+		} else if k == "duration" {
+			stats.Duration = value
+		}
+	}
+	return stats
+}
+
+// Decode key for fop and stat. keys starting with -1 belong to cumulative stat
+// Eg: -1-12-maxlatency fop : 12        stat: maxlatency
+// -1-duration          fop : ""        stat: duration
+func decodeCumulativeKey(key string) (string, string) {
+	var fop string
+	var k string
+	s := strings.Split(key, "-")
+	if len(s) == 4 {
+		k = s[len(s)-1]
+		index, _ := strconv.Atoi(s[2])
+		fop = fops[index]
+	} else if len(s) == 3 {
+		k = s[len(s)-1]
+		fop = ""
+	}
+	return fop, k
+}
+
+// Decode key for fop and stat for keys belonging to interval stat
+// Eg: 12-12-maxlatency  IntervalNo.: 12  fop : 12        stat: maxlatency
+// 12-duration           IntervalNo.: 12  fop : ""        stat: duration
+func decodeIntervalKey(key string) (string, string) {
+	var fop string
+	var k string
+	s := strings.Split(key, "-")
+	if len(s) == 3 {
```

Comment:
can't len be greater than 3?

Reply:
@atinmu I haven't come across any case yet where the len after splitting is other than 2 or 3 because as mentioned in the comments above this method there are 3 parts that make up a key of the profile info output i.e. interval no., fop number, stat name. 
In cumulative stats since we use "-1" to denote cumulative stat, on doing a split by "-"  we get 4 and sometimes 3 parts (depending on the stat).